### PR TITLE
logs: Introduce inline help

### LINKED ISSUE #13937: Inline help for logs filtering
Page: Logs

Spinning out of [PR #13905 ](https://github.com/cockpit-project/cockpit/pull/13905)

The syntax and power of the filter bar (regex etc.) would benefit from some inline documentation.
On top of that, @larskarlitski [suggested](https://github.com/cockpit-project/cockpit/pull/13905#issuecomment-614566979) we should have hints for `journalctl`. Well, actually, he suggested we should use the journalctl syntax 1:1, but that seems to bring it's own set of problems, so instead we can go for translating to journalctl output in some kind of documentation.

mockup

### PR DESCRIPTION
**Pull request by marusak**
This is rough implementation of inline help for logs. I though I would post it here so @andreasn can take a look and let me know what he thinks about it in this early stage. It contains all the info I think should be there, but I would appreciate help with wording it correctly :)
(and yes, I'll do in the meantime the journal command)
inlinehelp

Fixes #13937

### CODE REVIEW

**Comment by andreasn** [comment]
Great start!
It feels somewhat very big, so I wonder if we can try to cut something, since a deep-dive will be available under the _Learn more_ link
If I had to cut something, it would be the description of the quantifiers. I'm thinking you'll learn as you go by using the controls and see how they affect the text input.

For how to structure the text itself, I think Docs might be able to help us.

**Comment by marusak** [comment]
@andreasn made it more compact, implemented showing of the actual command and made it possible to copy into clipboard. WDYT?

newinlinehelp

**Comment by garrett** [comment]
It's promising!

There needs to be space above the headings. (Why isn't there already space above the headings? Hmm; weird...) Plus, the contents and headings should be aligned to the left. And the whole popup should have more padding inside.

Are you using completely custom styling or relying on PF4? There are enough differences that it feels like custom styling.

Also, there's no explanation what `journalctl`... is.

**Comment by garrett** [comment]
This should be a PF4 input group popover:

- Input group popover (React): https://www.patternfly.org/v4/documentation/react/components/inputgroup#with-popover (this is specifically what it should look and act like)

Here are the popover docs:

- Popover (React): https://www.patternfly.org/v4/documentation/react/components/popover
- Popover (HTML): https://www.patternfly.org/v4/documentation/core/components/popover#component-title

**Comment by marusak** [comment]
> Are you using completely custom styling or relying on PF4? 

Logs page does not include any PF4 and I don't think I want to introduce it in this step. It would complicate things a lot and would made the page much bigger (we would need to pull the whole PF4 styles). Now it is just pure html + css.

> Also, there's no explanation what journalctl... is.

Hmm, yeah. I want to keep it as small as possible (now it is already too big) and thought that we don't need to explain that this is command that generates the same output.

**Comment by andreasn** [comment]
I think the popup is a big, so I took a stab at simplifying the text a little.
The main idea I want to communicate is "look at how extensive this is" without going into detail about all the syntax, but leave that to the full documentation.
`priority` comes first, since that one is already in the text field by default, so the user can get a sense of how these work in action by looking at what's already in the text entry.

```
The text entry allows searching the full content of the log entrys. It accepts strings such as:
- qualifiers, E.g 'priority', 'identifier', 'services'
- log fields, E.g '_EXE=/usr/bin/python'
- any free-form string, can be combined using regular expressions
Learn more
```

Screenshot from 2020-05-26 13-03-50

**Comment by martinpitt** [comment]
@marusak: You marked this as release-blocker. If the design is finished now, can you please add a little test and run them? Thanks!

**Comment by marusak** [comment]
Current version:
new-log-inline-help

Also wrote tests. There is one remaining problem, that on smaller screens when labels start to wrap into columns, search does not span:
mobiledowsnotspan

I'll try to figure something out (spinning tests in the meantime), but if I cannot find any good solution and @garrett is out today, then I guess lets postpone this.

**Comment by marusak** [comment]
Ah typical, once I wrote comment I found the solution. Fixed.

File: pkg/systemd/logs.html [hunk 1]
```diff
@@ -67,7 +67,27 @@
...
             <label for="journal-grep" translate>Text</label>
-            <input id="journal-grep" class="form-control" type="text" translate="placeholder" placeholder="Type to filter">
+            <span class="filter-group">
+                <input id="journal-grep" class="form-control" type="text" translate="placeholder" placeholder="Type to filter">
+                <div id="logs-help" class="dropdown">
+                  <a role="link" tabindex="0" id="logs-help-toggle" class="dropdown-toggle dropdown-form-control" data-toggle="dropdown">
+                    <span class="pficon pficon-help"></span>
+                  </a>
+                  <div id="logs-help-menu" class="dropdown-menu">
+                    <span translate>The text entry allows searching the full content of the log entries. It accepts strings such as:</span>
```
Comment (martinpitt) [inline pkg/systemd/logs.html:77]: This is rather long, and translations won't make that better. Can we trim this a little bit? Maybe "Filter the logs with a combination of terms:"?

File: pkg/systemd/logs.html [hunk 2]
```diff
@@ -67,7 +67,27 @@
...
+                    <ul>
+                        <li><span translate>qualifiers</span>, <span translate>E.g</span> 'priority', 'identifier', 'services'</li>
```
Comment (martinpitt) [inline pkg/systemd/logs.html:79]: s/E.g/e.g./. I also think this should mention the colon. Also, it seems it's "service:", not ":service*s*"

File: pkg/systemd/logs.html [hunk 3]
```diff
@@ -67,7 +67,27 @@
...
+                        <li><span translate>log fields</span>, <span translate>E.g</span> '_EXE=/usr/bin/python'</li>
+                        <li translate>any free-form string, can be combined using regular expressions</li>
```
Comment (martinpitt) [inline pkg/systemd/logs.html:81]: I'm not sure how regular expressions  can combine things. This means that terms which are not "qualifier:" or "FIELD=" are interepreted as RE, right? So maybe just "a regular expression", and the introduction above covers the "combination" part.

File: pkg/systemd/logs.html [hunk 4]
```diff
@@ -67,7 +67,27 @@
...
+                    </ul>
+                    <a href="https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/managing_systems_using_the_rhel_8_web_console/reviewing-logs_system-management-using-the-rhel-8-web-console#filtering-logs-in-the-web-console_reviewing-logs" target="blank" rel="noopener noreferrer"><span translate>Learn more</span><i class="fa fa-external-link fa-xs"></i></a>
```
Comment (martinpitt) [inline pkg/systemd/logs.html:83]: Could we also link to the [journalctl manpage](https://www.freedesktop.org/software/systemd/man/journalctl.html)?

File: pkg/systemd/logs.html [hunk 5]
```diff
@@ -67,7 +67,27 @@
...
+                    <span class="journal-cmd">
+                        <pre id="journal-query"></pre>
+                        <button class="copy-cmd" id="journal-cmd-copy"><i class="fa fa-clipboard fa-xs"></i></button>
```
Comment (martinpitt) [inline pkg/systemd/logs.html:86]: I find it a bit irritating that without fields filters the command always ends in `--`. This looks like it's cut off. It's really not (it's just not getting any additional specifiers), but maybe we can only append an option separator if there are actually some extra search terms?

**Comment by marusak** [comment]
Adjusting to @martinpitt review:
ininehelp21

**Comment by marusak** [comment]
(test is not working well, I'll debug it locally and then push these changes)

**Comment by marusak** [comment]
Had to add special case for debian-stable

**Comment by garrett** [comment]
It still doesn't look PF4ish. Just because we might not be using PF4 on the page doesn't mean it should look wildly different.

First, let's navigate to the element. The CSS in the devtools is:

```scss
.pf-c-popover__content {
    position: relative;
    padding: var(--pf-c-popover__content--PaddingTop) var(--pf-c-popover__content--PaddingRight) var(--pf-c-popover__content--PaddingBottom) var(--pf-c-popover__content--PaddingLeft);
    background-color: var(--pf-c-popover__content--BackgroundColor);
}
```

According to devtools layout, the padding is 32px.

Screenshot from 2020-05-27 10-11-55

This means it's 2rem, as 1rem is set to 16px. And, sure enough, looking at the CSS, we can see the PF4 popovers are spaced with consistent padding, including `--pf-c-popover__content--PaddingTop`, which Firefox says equates to `--pf-global--spacer--xl`.

The shadow is set with `box-shadow: var(--pf-c-popover--BoxShadow)`... but since it's `pf-c-`, that's a component. We aren't using PF4 directly, so we'd need to use the global variable that it references. (`--pf-c-` variables are almost always aliases to `--pf-global-` variables). In this case, hovering the variable in Firefox shows what it is set to. In this case, it's `--pf-global--BoxShadow--md`.

Headings should have standard sizes too. We'll use `h3` as an example here. (I'm not sure what would be the correct heading here without looking at the context. PF4 uses `h1`s in the example. Perhaps we should use that? \*shrug\*)

Again, let's look up the `pf-c` variables and map them to `pf-global`:
- `--pf-c-title--m-xl--FontSize`: `--pf-global--FontSize--xl`
- `--pf-c-title--m-xl--FontWeight`: `--pf-global--FontWeight--normal`
- `--pf-c-title--m-xl--LineHeight`: `--pf-global--LineHeight--md`

_Therefore_, the popover, to resemble PF4 and use PF4 variables (which we do import into every page regardless if PF4 is actively being used), should look something like:

```scss
.ct-popover-example {
  box-shadow: var(--pf-global--BoxShadow--md);
  padding: var(--pf-global--spacer--xl);

  .ct-popover-example-heading {
    font-size: var(--pf-global--FontSize--xl);
    font-weight: var(--pf-global--FontWeight--normal);
    line-height: var(--pf-global--LineHeight--md);
  }
}
```

And that's how you can reverse out what PF4 is doing and mimic the style... all with PF4-based variables, so it should be resilient. If/when PatternFly changes the variables in the future, the will update here too, and it'll still fit in. (The most likely one to change in the upcoming future is the shadow.)

**Comment by garrett** [comment]
I still want to work on the strings.

Also, the command should wrap, not scroll. It probably needs something like `white-space: pre-wrap` (or one of the other ones, depending...). Docs: https://developer.mozilla.org/en-US/docs/Web/CSS/white-space

**Comment by marusak** [comment]
> It still doesn't look PF4ish

it looks like other inline helps we have. I did exactly the same thing you now descirbed to make sure it looks like other inline helps:
this

> Also, the command probably should wrap, not scroll.

I don't like this idea as this is kind of learning/discoverablity tool - you want to see what is there and not just still see the same begining and need to copy it, then go somewhere where you can actually paste it to see, what it contains. For example if I would query specific, service, I still would only see `journalctl --no-tail --priority=err --reverse`  which does not seem right.

**Comment by marusak** [comment]
> and I think tweaking the design can be a follow-up matter.

Agree, lets not block this on whether padding should be `1rem` and look like other inline helps or `2rem` and look like PF4 popups. It is more important that it gets released and can be safely shipped into 8.3 together with the search box.

**Comment by garrett** [comment]
> it looks like other inline helps we have. I did exactly the same thing you now descirbed to make sure it looks like other inline helps:

...but that isn't _inline_ help. That's a screenshot of the global help. The inline help screenshot is far above and is a popover. The global help is a dropdown (not a popover).

**Comment by andreasn** [comment]
Since this is merged now, can we split that discussion on styling into a new issue?

**Pull request merged**